server/neptune/workflows/internal/deploy/revision/queue/worker.go

```diff
@@ -45,39 +62,106 @@ func (w *Worker) Work(ctx workflow.Context) {
 		w.state = CompleteWorkerState
 	}()
 
-	var latestDeployment *deployment.Info
-	for {
-		if w.Queue.IsEmpty() {
-			w.state = WaitingWorkerState
-		}
-
-		err := workflow.Await(ctx, func() bool {
-			return !w.Queue.IsEmpty()
-		})
+	var previousDeployment *deployment.Info
+	var currentAction actionType
+	callback := func(f workflow.Future) {
+		err := f.Get(ctx, nil)
 
 		if temporal.IsCanceledError(err) {
-			logger.Info(ctx, "Received cancelled signal, worker is shutting down")
+			currentAction = canceled
 			return
 		}
 
 		if err != nil {
 			logger.Error(ctx, fmt.Sprintf("Unknown error %s, worker is shutting down", err.Error()))
+			currentAction = canceled
 			return
 		}
 
-		w.state = WorkingWorkerState
+		currentAction = process
+	}
+	selector := workflow.NewSelector(ctx)
+	selector.AddFuture(w.awaitWork(ctx), callback)
```

Comment:
nit: can we add this future callback at the start of the for loop to avoid having to add it again at the end?

Reply:
no we can't its a future so every time one is done you need to add another